libbeat/publisher/queue/diskqueue/deleter_loop.go

```diff
@@ -0,0 +1,94 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package diskqueue
+
+import (
+	"errors"
+	"os"
+	"time"
+)
+
+type deleterLoop struct {
+	// The settings for the queue that created this loop.
+	settings *Settings
+
+	// When one or more segments are ready to delete, they are sent to
+	// requestChan. At most one deleteRequest may be outstanding at any time.
+	requestChan chan deleterLoopRequest
+
+	// When a request has been completely processed, a response is sent on
+	// responseChan. If at least one deletion was successful, the response
+	// is sent immediately. Otherwise, the deleter loop delays for
+	// queueSettings.RetryWriteInterval before returning, so timed retries
+	// don't have to be handled by the core loop.
+	responseChan chan deleterLoopResponse
+}
+
+type deleterLoopRequest struct {
+	segments []*queueSegment
+}
+
+type deleterLoopResponse struct {
+	// The queue segments that were successfully deleted.
+	deleted map[*queueSegment]bool
+
+	// Errors
+	errors []error
```

Comment:
I suspect that particular errors are linked with particular queueSegments. WDYT about an array with "delete statuses" instead of map? The status may contain an error.

Reply:
I like that, it's more consistent with how the writer loop behaves too. Done.